PHP 8 Support

utopiabound opened the pull request:
Support for PHP4 and PHP5 is dropped
Convert while (list($key, $var) = each($array)) to foreach()
Convert any array {} lookups to []
Upgrade libchart to 1.3 (final)
Upgrade Snoopy to 2.0.0 (final)
Upgrade PHPMailer to 5.2.28 (final before 6.x rewrite)
Upgrade PHPThumb to 1.7.17 (
Remove earlier libchart versions.
Fix many uninitialized values (fixes warnings in logs)

utopiabound committed:
PHP 8 Support

Support for PHP4 and PHP5 is dropped
Convert while (list($key, $var) = each($array)) to foreach()
Convert any array {} lookups to []
Upgrade libchart to 1.3 (final)
Upgrade Snoopy to 2.0.0 (final)
Upgrade PHPMailer to 5.2.28 (final before 6.x rewrite)
Upgrade PHPThumb to 1.7.17 (
Remove earlier libchart versions.
Fix many uninitialized values (fixes warnings in logs)

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

utopiabound committed:
Fix item_input

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

utopiabound committed:
Fix typo

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

NicFord1 commented on lib/AdminAjaxJobs.class.php:
```diff
@@ -13,7 +13,7 @@ class AdminAjaxJobs {
 	var $_debug;
 
-	function AdminAjaxJobs($id, $job, $batchlimit) {
+	function __construct($id, $job, $batchlimit) {
```
This creates a PHP fatal error in relation to file: ./admin/item_cache/ItemCacheAjaxJobs.class.php

> [17-Oct-2021 11:37:50 America/Boise] PHP Fatal error:  Uncaught Error: Call to undefined method AdminAjaxJobs::AdminAjaxJobs() in ./admin/item_cache/ItemCacheAjaxJobs.class.php:7
Stack trace:
#0 ./admin/item_cache/ajaxjobs.php(6): ItemCacheAjaxJobs->ItemCacheAjaxJobs(NULL)
#.1 ./admin.php(58): include_once('...')
#.2 {main}
  thrown in ./admin/item_cache/ItemCacheAjaxJobs.class.php on line 7

To fix, line 7 of ./admin/item_cache/ItemCacheAjaxJobs.class.php should be updated too:
`		parent::__construct(get_class($this), $job, 10);`

NicFord1 commented on lib/logging.php:
```diff
@@ -80,64 +80,64 @@ function fget_tokenised_log_entry(&$file) {
 					}
 					
-					$column .= $line {$i};
+					$column .= $line[$id];
```
I believe this is a typo, should be $i not $id

NicFord1 reviewed (commented):
Came across two code issues.

utopiabound committed:
Run php lint and adress review comments

phplint: "php -l" on each php file
A few more warnings, and bugs squashed.

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

NicFord1 commented on lib/widgets.php:
```diff
@@ -98,16 +98,15 @@ function custom_select($name, $lookup_results, $display_mask, $size = 1, $value
 	
 	$value_found = FALSE;
-	while ( $lookup_r = each ( $lookup_results ) ) {
+	foreach ( $lookup_results as $lookup_r ) {
```
Breaks functionality as $lookup_r is no longer an associative array.

Broken functionality examples can be observed on:
- siteurl/admin.php?type=s_file_type with the content group column
- siteurl/admin.php?type=config&group_id=site for the default theme selection.

Working code below:
```
	foreach ( $lookup_results as $lookup_r_key => $lookup_r_value ) {
		$lookup_r = ["key" => $lookup_r_key, "value" => $lookup_r_value];
```

utopiabound replied:
This code works for me with php 8.0.11.  What issue are you seeing?

*Update:* Oh, I see it.  The "Default Theme" selector.

utopiabound committed:
Address review comments, and fix other bugs

Remove some unused "$key" variables in foreach statements.
Use ?? operator instead of checking isset() in some places.

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

utopiabound committed:
Fix typos

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

utopiabound committed:
More warnings silenced

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

utopiabound committed:
Fix more warnings and uninitialized values

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

utopiabound committed:
Fix incorrect const, and warnings

borrow.php: Fix incorrect constant name

Fix warnings of uninitialized variables.

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

utopiabound committed:
Fix bug in item input

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>

utopiabound committed:
Fix admin attribute changes

Also more warnings

Signed-off-by: Nathaniel Clark <Nathaniel.Clark@misrule.us>
